lib/ASTGen/Sources/ASTGen/Exprs.swift

```diff
@@ -1,6 +1,57 @@
 import ASTBridging
 import SwiftDiagnostics
-import SwiftSyntax
+@_spi(ExperimentalLanguageFeatures) import SwiftSyntax
+
+/// Check if an `ExprSyntax` can be generated using ASTGen.
+///
+/// If all the expression nodes that shares the first token are migrated,
+/// returns true. For example, given
+///   ```
+///   foo.bar({ $0 + 1 }) + 2
+///   ```
+/// `foo` token is the first token of all `SequenceExpr`, `FunctionCallExpr`,
+/// `MemberAccessExpr`, and `DeclReferenceExpr`. All these expression kinds must
+/// be migrated to handle it in ASTGen. Because the fallback
+/// `generateWithLegacy(_:)` only receives the parser position, it eagerly
+/// parses the outer expressions.
+func isExprMigrated(_ node: ExprSyntax) -> Bool {
+  var current: Syntax = Syntax(node)
+  if let firstToken = node.firstToken(viewMode: .sourceAccurate) {
+    current = firstToken.parent!
+  }
+  while true {
+    switch current.kind {
+    case // Known implemented kinds.
+        .closureExpr, .functionCallExpr, .declReferenceExpr, .memberAccessExpr,
+        .tupleExpr:
+      break
+    case // Known unimplemented kinds.
+        .arrayExpr, .arrowExpr, .asExpr, .assignmentExpr, .awaitExpr,
+        .binaryOperatorExpr, .booleanLiteralExpr, .borrowExpr, .canImportExpr,
+        .canImportVersionInfo, .dictionaryExpr, .discardAssignmentExpr, .doExpr,
+        .editorPlaceholderExpr, .floatLiteralExpr, .forceUnwrapExpr, .inOutExpr,
+        .infixOperatorExpr, .integerLiteralExpr, .isExpr, .keyPathExpr,
+        .macroExpansionExpr, .consumeExpr, .copyExpr, .nilLiteralExpr,
+        .optionalChainingExpr, .packElementExpr, .packExpansionExpr,
+        .postfixIfConfigExpr, .postfixOperatorExpr, .prefixOperatorExpr,
+        .regexLiteralExpr, .sequenceExpr, .genericSpecializationExpr,
+        .stringLiteralExpr, .simpleStringLiteralExpr, .subscriptCallExpr,
+        .superExpr, .switchExpr, .ternaryExpr, .tryExpr, .typeExpr,
+        .unresolvedAsExpr, .unresolvedIsExpr, .patternExpr,
+        .unresolvedTernaryExpr:
+      return false
+    case // Unknown expr kinds.
+      _ where current.is(ExprSyntax.self):
+      return false
+    default:
+      break;
+    }
+    if current.id == node.id {
+      return true
+    }
+    current = current.parent!
```

Comment:
I don’t think `current.parent` needs to be non-nil if the entire syntax tree starts with an `ExprSyntax`. I think we should return `true` if there is no more parent.

Reply:
This is walking up the parents from the first token of `node`. `.parent` must exist if `current` is not `node`

Reply:
Oh, I see. Could you add a comment that contains exactly what you said in the reply?